tensorflow/python/framework/ops_test.py

```diff
@@ -148,6 +150,24 @@ def testShapeFunctionError(self):
           r"\(op: 'Add(V2)?'\) with input shapes: \[1,2,3\], \[4,5,6\]."):
         _ = a + b
 
+  def testNumpyArray(self):
+    with context.graph_mode():
+      x = array_ops.ones((3, 4), name="test_ones")
+
+    with self.assertRaisesRegexp(NotImplementedError,
+                                 r"Cannot convert a symbolic.+test_ones"):
+      np.array(x)
```

Comment:
i'm not sure this is a backwards compatible change.  right now np.array(tensor) just wraps the tensor as a scalar in an array.  breaking this behavior is questionable.  @alextp wdyt?

Reply:
I think we should run a large-scale integration test and look at the failures. Overall I think this change is an improvement though as with mixed eager / graph code it's a footgun.

Reply:
So it seems that overriding `__array__` fundamentally changes the behavior of all interactions with numpy's ufuncs. For instance given the current behavior for the following
```
x = np.array(1.)
y = tf.ones(())

x / y
y / x
np.divide(x, y)
```
is that all three divisions call the tf divide function and return a single Tensor. However, once `__array__` is overridden to produce an error then `np.divide` fails. This will be true for all ufuncs. (no more `np.sum(my_tensorflow_tensor)`) I think in order to both disallow `np.array(my_tensorflow_tensor)` and properly support numpy dispatch we will have to implement either the `ufunc` or `array_function` protocols for Tensors which is a much more significant undertaking.

This also suggests that we should have better unit tests for numpy compatibility in tensorflow, since this was only caught by the large-scale integration test.